### ISSUE TITLE
Default value polyfill

### ISSUE THREAD

**Issue by markogresak** [babel/babel#2054]
I was checking how babel replaces function's default values and I'd like explanation for decisions behind this kind of polyfill:

Source:
``` js
function  example(a, b = 2, c = 'c') {}
```
Output:
``` js
function example(a) {
  var b = arguments.length <= 1 || arguments[1] === undefined ? 2 : arguments[1];
  var c = arguments.length <= 2 || arguments[2] === undefined ? 'c' : arguments[2];
}
```

Is there any other reason behind checking the length first other than the performance?

As far as I know this should work even in IE6 and I'm aware that not checking the length is approx. 20% slower, but this is still not significant, it's barely noticeable at 3M samples (~15ms in chrome).
On the other hand, there is 21 characters less or 21B less (assuming utf-8 encoding) for each optional argument polyfill. This might be more significant, especially in large apps, especially to those loading it over slow internet connection.

**Comment by loganfsmyth** [comment]
Do you think it will be that noticeable on the whole? I'd assume as long as the code is gzipped, the difference would be negligible.

**Comment by sebmck** [comment]
The extra code size is worth the performance. You should be able to trust Babel to output the most perfomant code possible. As Logan pointed out, the code size is negligible when gripped, if you're pedantic about the size of your code to that degree then you shouldn't be using Babel or any JavaScript preprocessor.

**Comment by markogresak** [comment]
Forgot to factor in the gzipping, the LZ algorithm will reduce these lines to almost nothing.

@sebmck I do trust babel and I wouldn't change it unless would make a tremendous change (i.e. never). This was more of me asking of explanation for this decision, since I haven't used this style before and I was curious of why there was this extra length check.

**Comment by sebmck** [comment]
Sure, no problem. Will provide some references to issues when I can that lead up to the current behaviour.

**Comment by sebmck** [comment]
Related discussion available in #1845 and [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1131996) Mozilla bug report.

**Comment by markogresak** [comment]
Thanks for the links. I noticed that Chrome and Safari are significantly faster than Firefox, it's actually like this with most of the array operations.
